Add support for chunking of blobs, using a variant of BLAKE3

EdSchouten opened the pull request:
Buildbarn has invested heavily in using virtual file systems. Both on the worker and client side it's possible to lazily fault in data from the CAS. As Buildbarn implements checksum verification where needed, randomly accessing large files may be slow. To address this, this change adds support for composing and decomposing CAS objects, using newly added ConcatenateBlobs() and SplitBlobs() operations.

If implemented naively (e.g., using SHA-256), these operations would not be verifiable. To rephrase: when merely given the checksum of smaller objects, there is no way to obtain that of its concatenated version. This is why at the same time, this change adds a new digest function that closely resembles BLAKE3. BLAKE3 is based on a binary Merkle tree, meaning that it's possible to efficiently concatenate and split objects at the 2^k boundary (where k > 10).

With these new operations present, there is no true need to use the Bytestream protocol any longer. Writes can be performed by uploading smaller parts through BatchUpdateBlobs(), followed by calling ConcatenateBlobs(). Conversely, reads of large objects can be performed by calling SplitBlobs() and downloading individual parts through BatchReadBlobs(). At no point is integrity compromised, as callers of SplitBlobs() can validate the resulting tree nodes against the original digests.

One feature of BLAKE3 is that its hashes are variable length. Though this is nice to have (allowing users to make size/security tradeoffs), it does mean that the length of the hash cannot be used to infer the digest function used. This has already become an issue with MD5 vs MURMUR3. To solve that, we extend all operations that work with digests to take a digest function explicitly. For compatibility, we allow this to be UNKNOWN for all existing digest functions.

moroten commented on build/bazel/remote/execution/v2/remote_execution.proto:
```diff
@@ -1839,6 +2010,45 @@ message CacheCapabilities {
+  // not exceed `max_batch_total_size_bytes`.
+  int32 blake3concat_max_upload_size_bytes = 8;
+
+  // The minimum size of blobs that can be created by calling
+  // [SplitBlobs][build.bazel.remote.execution.v2.ContentAddressableStorage.SplitBlobs]
+  // against a blob that uses the BLAKE3CONCAT digest function,
+  // disregarding the blob containing trailing data.
+  //
+  // If supported, this field MUST have value 2^k, where k > 10. It may
+  // not exceed `max_batch_total_size_bytes`.
+  int32 blake3concat_min_split_size_bytes = 9;
```
Will the server be able to have an efficient implementation if the API is called with different chunk sizes? Would there be an "optimal" value for a static chunk size to reduce the risk of inefficient client-server combinations?

EdSchouten replied:
I think that in the case of Buildbarn I will just set `blake3concat_max_upload_size_bytes == blake3concat_min_split_size_bytes`. Then all clients/servers will be in full agreement on what the chunk size is.

The reason I kept them apart, is because I tried to keep the upload & download paths separated. I can imagine that if someone operating a cluster discovers that they made a bad choice regarding chunk size, that having these separate will make it easier to migrate from one chunk size to the other if needed.

With regards to a one size fits all optimal chunk size, I'm not sure whether such a value exists. I guess it really depends on bandwidth vs. latency. If you're bandwidth constrained, then smaller chunks is better. Conversely, if latency is high, it may be desirable to set the chunk size higher, so that you need to call into ConcatenateBlobs() and SplitBlobs() less frequently.

moroten reviewed (commented):
This looks like a promising way to get this feature into v2 instead of waiting for v3. Great work!

EdSchouten commented on build/bazel/remote/execution/v2/remote_execution.proto:
```diff
@@ -1600,6 +1662,78 @@ message BatchUpdateBlobsResponse {
+  // The digests of the large blobs to split into smaller blobs.
+  repeated Digest digests = 3;
+
+  // The size of the non-trailing blobs to create. It must be less than
+  // the sizes stored in `digests`. Furthermore, for BLAKE3CONCAT must
+  // be a power of 2.
+  int64 split_size_bytes = 4;
```
It depends. If the client wants really granular access to the object with as much deduplication as possible, it can use `blake3concat_min_split_size_bytes`. If it simply wants to download the object in its entirety without caring too much about about deduplication, it can use the highest power of two below `max_batch_total_size`.

EdSchouten commented on build/bazel/remote/execution/v2/remote_execution.proto:
```diff
@@ -381,6 +389,25 @@ service ContentAddressableStorage {
 
+  // Create blobs by concatenating multiple smaller blobs.
+  //
+  // This method is only supported for digest function BLAKE3CONCAT. For
+  // other digest functions new blobs can only be created by uploading
+  // them in their entirety.
```
Let me improve the phrasing here. The reason things are phrased the way they are, is because SplitBlobs() is documented as a call that *creates* blobs for you. And for the other hashing functions that can only be done by uploading.

But because this call is in practice used as a prerequisite for downloading objects, let me document it as such.

EdSchouten commented on build/bazel/remote/execution/v2/remote_execution.proto:
```diff
@@ -381,6 +389,25 @@ service ContentAddressableStorage {
+  // other digest functions new blobs can only be created by uploading
+  // them in their entirety.
+  //
+  // This method may be called recursively to split very large objects.
```
Imagine the pathological case where you want to upload a 1 PiB (2^50 bytes) file into the CAS. If `blake3concat_max_upload_size_bytes` is 1 MiB (2^20 bytes), you would need to call ConcatenateBlobs() with `chaining_values` being 64 * 2^30 bytes in size. This is obviously not possible, as it strongly exceeds the maximum gRPC message body.

What this sentence means to say is that you can construct such blobs by first issuing 2^15 calls to ConcatenateBlobs() to construct CAS objects that are 32 GiB (2^35 bytes) each. After you're done issuing those, you can call ConcatenateBlobs() once more to create the 1 PiB object.

EdSchouten commented on build/bazel/remote/execution/v2/remote_execution.proto:
```diff
@@ -1600,6 +1662,78 @@ message BatchUpdateBlobsResponse {
+    // bytes of data, where 0 < l <= 2^k. The server may infer the
+    // values of k and l using `digest.size_bytes` and the length of
+    // this field.
+    bytes chaining_values = 2;
```
So initially I actually had a `repeated fixed32` here, so that the 32-bit integers would be stored as little-endian values directly. The reason I picked `bytes` here is to keep this generic enough, so that it may work for other hashing algorithms as well.

The documentation that I gave above purely reflects how this field is used in the case of BLAKE3CONCAT. Other hashing algorithms are free to use other layouts/representations that are not necessarily record based.

EdSchouten replied:
No longer applies, as we now use string hashes here.

EdSchouten commented:
As #235 and #236 are in my opinion close to a state in which they can be merged, I have gone ahead and reimplemented this PR on top of #235. Changes to the previous version are as follows:

- As suggested by @EricBurnett + @bergsieker, I have moved the {Concatenate,Split}Blobs() capabilities into a separate message. In theory you could use it in combination with any digest function. The downside of using anything other than SHA256TREE is obviously that client/server-side validation of these requests is either impossible or prohibitively expensive.
- Related to the above, {Concatenate,Split}Blobs() now take/return hashes of small objects, instead of BLAKE3-style chaining values. This 
- FindMissingBlobsRequest now also has a new `split_sizes_bytes` field. This permits clients to more efficiently increase the lifetime of objects returned by SplitBlobs().

PTAL, ignoring the first commit in this PR. That one is part of #235.

stagnation commented on build/bazel/remote/execution/v2/remote_execution.proto:
```diff
@@ -1600,6 +1693,76 @@ message BatchUpdateBlobsResponse {
+    // `split_size_bytes`. The final element corresponds to the the
+    // non-empty blob containing trailing data, having size
+    // `digest.size_bytes % split_size_bytes`.
+    string small_hashes = 2;
```
Should these be `repeated`?

EdSchouten replied:
Yes! I omitted this by accident when I changed the field from `bytes` to a plain list of hashes.

tylerwilliams commented:
I think this idea is neat! I'm excited about combining this with the idea of transmitting variable size chunks.

I'd like to propose a slight modification: instead of `repeated string small_hashes = 2`, if we repeated Digest messages, we could allow for multiple digests of different sizes to be joined into a contiguous chunk. It's also a more natural fit -- small chunks that have been uploaded to the CAS already return a digest.

Separately, what do you think about returning a ConcatenateBlobsResponse from the ConcatenateBlobs call? Though the response matches BatchUpdateBlobsResponse today, it may not in the future, and it's much easier to change this now before clients are using it.

Finally, would it make sense to remove the chunk size in SplitBlobsRequest and let the server decide this value?
